modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/SubjectHierarchyValidator.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.uhndata.cards.forms.internal;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+
+import org.apache.jackrabbit.oak.api.CommitFailedException;
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.spi.commit.DefaultValidator;
+import org.apache.jackrabbit.oak.spi.commit.Validator;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+
+import io.uhndata.cards.subjects.api.SubjectTypeUtils;
+import io.uhndata.cards.subjects.api.SubjectUtils;
+
+/**
+ * A {@link Validator} that ensures that a child Subject has a parent of required type.
+ *
+ * @version $Id$
+ */
+public class SubjectHierarchyValidator extends DefaultValidator
+{
+
+    private final SubjectTypeUtils subjectTypeUtils;
+    private ThreadLocal<LinkedList<NodeState>> parentNodes = ThreadLocal.withInitial(LinkedList::new);
+
+    public SubjectHierarchyValidator(SubjectTypeUtils subjectTypeUtils)
+    {
+        this.subjectTypeUtils = subjectTypeUtils;
+    }
+
+    @Override
+    public Validator childNodeAdded(final String name, final NodeState after) throws CommitFailedException
+    {
+        this.parentNodes.get().add(after);
+        // Get the type of this node. Return immediately if it's not a cards:Subject node
+        final String childNodeType = after.getName("jcr:primaryType");
+        if (!(SubjectUtils.SUBJECT_NODETYPE).equals(childNodeType)) {
+            return new SubjectHierarchyValidator(this.subjectTypeUtils);
+        }
+
+        return isFormValid(after);
+    }
+
+    public Validator isFormValid(NodeState after) throws CommitFailedException
```

Comment:
The convention is that any method called `is...` returns a boolean. A better name would be `check...`.

Since the same `SubjectHierarchyValidator` is returned in any way other than an exception, it should be returned not from within this method, but from `childNodeAdded`, so nothing needs to be returned other than the occasional exception.

This doesn't check forms, but subjects, more specifically the subject's hierarchy.

Also, the method should not be public, everything should only be as exposed as needed. A better signature is:

```suggestion
    private void checkSubjectHierarchy(NodeState subject) throws CommitFailedException
```